engine/apps/alerts/tasks/notify_user.py

```diff
@@ -418,17 +468,144 @@ def perform_notification(log_record_pk):
 
 
 @shared_dedicated_queue_retry_task(
-    autoretry_for=(Exception,), retry_backoff=True, max_retries=0 if settings.DEBUG else None
+    autoretry_for=(Exception,), retry_backoff=True, max_retries=1 if settings.DEBUG else None
 )
-def send_user_notification_signal(log_record_pk):
-    start_time = time.time()
+def send_bundled_notification(user_notification_bundle_id):
+    from apps.alerts.models import AlertGroup, UserNotificationBundle
+    from apps.base.models import UserNotificationPolicy, UserNotificationPolicyLogRecord
+
+    task_logger.info(f"Start send_bundled_notification for user_notification_bundle {user_notification_bundle_id}")
+    with transaction.atomic():
+        try:
+            user_notification_bundle = UserNotificationBundle.objects.filter(
+                pk=user_notification_bundle_id
+            ).select_for_update()[0]
+        except IndexError:
+            task_logger.info(
+                f"user_notification_bundle {user_notification_bundle_id} doesn't exist. "
+                f"The user associated with this notification bundle may have been deleted."
+            )
+            return
+
+        if not compare_escalations(send_bundled_notification.request.id, user_notification_bundle.notification_task_id):
+            task_logger.info(
+                f"send_bundled_notification: notification_task_id mismatch. "
+                f"Duplication or non-active notification triggered. "
+                f"Active: {user_notification_bundle.notification_task_id}"
+            )
+            return
 
-    from apps.base.models import UserNotificationPolicyLogRecord
+        notifications = user_notification_bundle.notifications.filter(bundle_uuid__isnull=True)
+
+        alert_group_ids = [i.alert_group_id for i in notifications]
+        # get active alert groups to notify about
+        active_alert_groups = set(
+            AlertGroup.objects.filter(
+                id__in=alert_group_ids,
+                resolved=False,
+                acknowledged=False,
+                silenced=False,
+            ).values_list("id", flat=True)
+        )
+        created_lod_records = []
+        skip_notification_ids = []
+        perform_notifications = []
 
-    task_logger.debug(f"LOG RECORD PK: {log_record_pk}")
-    task_logger.debug(f"LOG RECORD LAST: {UserNotificationPolicyLogRecord.objects.last()}")
+        # create logs
+        for notification in notifications:
```

Comment:
Maybe it could be once for cycle, since we already iterated through notifications in line 500